### ISSUE TITLE
ComfyUI plugin for multi GPU inference

### ISSUE THREAD

**Issue by lanyan520** [Wan-Video/Wan2.1#250]
可以提供一个多GPU推理使用的ComfyUI插件么
Can you provide a ComfyUI plugin for multi GPU inference

**Comment by lanyan520** [comment]
![Image](https://github.com/user-attachments/assets/049f1c7e-c30d-458a-b9ba-0f619bd87f59)

import os
import subprocess
import comfy.utils
import folder_paths
import torch
import json
from datetime import datetime
from pathlib import Path

class WanI2V_Generator:
    """
    Wan2.1 多GPU视频生成节点
    文档：https://github.com/Wan-Video/Wan2.1
    """
    
    # 默认配置（根据官方仓库参数）
    DEFAULT_CONFIG = {
        "model_repo": "Wan-AI/Wan2.1-I2V-14B-720P-Diffusers",
        "hf_mirror": "https://hf-mirror.com",
        "num_frames": 24,
        "fps": 12,
        "guidance_scale": 7.5,
        "num_inference_steps": 50
    }

    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "image": ("IMAGE",),
                "prompt": ("STRING", {"multiline": True, "default": "A cinematic shot of..."}),
                "width": ("INT", {"default": 1280, "min": 512, "max": 4096}),
                "height": ("INT", {"default": 720, "min": 512, "max": 2160}),
                "num_gpus": ("INT", {"default": -1, "min": -1, "max": 64, "step": 1}),
            },
            "optional": {
                "advanced_config": ("CONFIG", {"default": None}),
                "seed": ("INT", {"default": -1}),
            }
        }

    RETURN_TYPES = ("IMAGE",)
    RETURN_NAMES = ("video_frames",)
    OUTPUT_IS_LIST = (True,)
    CATEGORY = "Video Generation/Wan2.1"
    FUNCTION = "generate"

    def generate(self, image, prompt, width, height, num_gpus, advanced_config=None, seed=-1):
        # ================== 初始化配置 ==================
        config = self.DEFAULT_CONFIG.copy()
        if advanced_config:
            config.update(advanced_config)
        
        # ================== 路径处理 ==================
        # 保存输入图像
        input_path = self.save_temp_image(image)
        
        # 准备输出目录
        output_dir = Path(folder_paths.get_output_directory())
        timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
        output_prefix = output_dir / f"wan2.1_{timestamp}"
        
        # ================== GPU配置 ==================
        available_gpus = torch.cuda.device_count()
        if num_gpus == -1:
            num_gpus = available_gpus
        num_gpus = min(num_gpus, available_gpus)
        
        # ================== 构建命令 ==================
        cmd = [
            "torchrun",
            f"--nproc_per_node={num_gpus}",
            "generate.py",
            "--hf-mirror", config["hf_mirror"],
            "--model_repo", config["model_repo"],
            "--image", str(input_path),
            "--prompt", prompt,
            "--width", str(width),
            "--height", str(height),
            "--num_frames", str(config["num_frames"]),
            "--fps", str(config["fps"]),
            "--guidance_scale", str(config["guidance_scale"]),
            "--num_inference_steps", str(config["num_inference_steps"]),
            "--output_prefix", str(output_prefix)
        ]
        
        if seed != -1:
            cmd += ["--seed", str(seed)]

        # ================== 执行生成 ==================
        try:
            self.run_command(cmd)
        except Exception as e:
            raise RuntimeError(f"生成失败: {str(e)}")

        # ================== 加载结果 ==================
        frame_paths = sorted(output_prefix.parent.glob(f"{output_prefix.name}_frame_*.png"))
        if not frame_paths:
            raise FileNotFoundError("未找到生成的帧序列")
            
        frames = [self.load_image(p) for p in frame_paths]
        return (frames,)

    def save_temp_image(self, image_tensor):
        """保存输入图像到临时目录"""
        temp_dir = Path(folder_paths.get_temp_directory()) / "wan2.1_inputs"
        temp_dir.mkdir(exist_ok=True)
        
        filename = f"input_{datetime.now().strftime('%Y%m%d%H%M%S')}.png"
        save_path = temp_dir / filename
        
        comfy.utils.save_image(image_tensor, str(save_path), quality=100)
        return save_path

    def load_image(self, path):
        """加载图像序列为张量"""
        img = comfy.utils.load_image(str(path))
        return torch.from_numpy(img).unsqueeze(0)

    def run_command(self, cmd):
        """执行命令行并捕获日志"""
        process = subprocess.Popen(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            bufsize=1,
            cwd="/path/to/Wan2.1"  # 修改为实际项目路径
        )

        # 实时输出日志
        while True:
            line = process.stdout.readline()
            if not line and process.poll() is not None:
                break
            if line:
                print(f"[Wan2.1] {line.strip()}")

        if process.returncode != 0:
            raise subprocess.CalledProcessError(process.returncode, cmd)

# ================== 插件注册 ==================
NODE_CLASS_MAPPINGS = {
    "Wan2.1_I2V_Generator": WanI2V_Generator
}

NODE_DISPLAY_NAME_MAPPINGS = {
    "Wan2.1_I2V_Generator": "🎬 Wan2.1 Video Generator (Multi-GPU)"
}

__all__ = ['NODE_CLASS_MAPPINGS', 'NODE_DISPLAY_NAME_MAPPINGS']

**Comment by lanyan520** [comment]
use ai create nodes.py ,but python main.py error:

Skip /mnt/comfyui/ComfyUI_0312/ComfyUI_Development/ComfyUI_Wan2.1/custom_nodes/Wan2I2V_Plugin module for custom nodes due to the lack of NODE_CLASS_MAPPINGS.

ComfyUI/
── custom_nodes/
│   └── Wan2I2V_Plugin/
│       ├── __init__.py
│       ├── nodes.py       # 本文件
│       └── utils/         # 辅助工具
└── Wan2.1/                # 从GitHub克隆的仓库
    ├── generate.py
    └── ...其他项目文件

**Comment by lcax200000** [comment]
class WanDistributedGenerator:
    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "task": (["t2v-14B", "t2i-14B", "i2v-14B"],),
                "prompt": ("STRING", {"default": "", "multiline": True}),
                "ckpt_dir": ("STRING", {"default": "models/Wan2.1-T2V-14B"}),
                "size": (["1280*720", "832*480", "480*832"],),
                "frame_num": ("INT", {"default": 81}),
                "nproc_per_node": ("INT", {"default": 8, "min": 1}),
                "ulysses_size": ("INT", {"default": 8, "min": 1}),
                "ring_size": ("INT", {"default": 1, "min": 1}),
                "sample_steps": ("INT", {"default": 50}),
                "guide_scale": ("FLOAT", {"default": 5.0}),
                "seed": ("INT", {"default": -1}),
                "t5_fsdp": ("BOOLEAN", {"default": True}),
                "dit_fsdp": ("BOOLEAN", {"default": True}),
                "save_dir": ("STRING", {"default": "outputs"}),
                "output_name": ("STRING", {"default": ""}),
            },
            "optional": {
                "input_image": ("IMAGE",),
            }
        }

    RETURN_TYPES = ("STRING",)
    RETURN_NAMES = ("output_path",)
    FUNCTION = "run_distributed"
    CATEGORY = "wan2.1"
    OUTPUT_NODE = True
    def run_distributed(self, task, prompt, ckpt_dir, size, frame_num, 
                        nproc_per_node, ulysses_size, ring_size,
                        sample_steps, guide_scale, seed, 
                        t5_fsdp, dit_fsdp, save_dir, output_name,
                        input_image=None):
        # 清理GPU内存
        import gc
        gc.collect()
        torch.cuda.empty_cache()
        
        # 创建日志目录
        logs_dir = os.path.join(save_dir, 'logs')
        os.makedirs(logs_dir, exist_ok=True)
        
        # 准备日志文件
        log_file_path = os.path.join(logs_dir, f"torchrun_log_{int(time.time())}.txt")
        
        # 构建torchrun参数
        generate_script = "/data/usp/Wan2.1/generate.py"
        torchrun_args = [
            "--nproc_per_node", str(nproc_per_node),
            "--master_addr", "localhost",
            "--master_port", str(29500 + random.randint(0, 1000)),
            generate_script,
            "--task", task,
            "--prompt", prompt,
            "--ckpt_dir", ckpt_dir,
            "--size", size,
            "--ulysses_size", str(ulysses_size),
            "--offload_model", "True",  # 启用模型卸载
            "--t5_cpu",  # 将T5模型放在CPU上
            "--sample_steps", str(sample_steps),
            "--sample_guide_scale", str(guide_scale),
            "--base_seed", str(seed),
            "--sample_solver", "unipc"
        ]
        
        # 添加可选参数
        if t5_fsdp:
            torchrun_args.append("--t5_fsdp")
        if dit_fsdp:
            torchrun_args.append("--dit_fsdp")
        if input_image:
            torchrun_args.extend(["--image", input_image])
            
        # 将命令保存到文件，方便调试
        params_file = os.path.join(logs_dir, f"params_{int(time.time())}.json")
        with open(params_file, 'w') as f:
            json.dump({
                "command": "torch.distributed.run",
                "args": torchrun_args,
                "time": datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            }, f, indent=2)
        
        print(f"启动分布式任务，命令已保存到: {params_file}")
        
        # 设置环境变量
        os.environ["PYTHONPATH"] = f"/data/usp/Wan2.1:{os.environ.get('PYTHONPATH', '')}"
        # 设置CUDA内存分配器
        os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "max_split_size_mb:128,garbage_collection_threshold:0.8"
        # 限制每个进程使用的GPU内存
        os.environ["CUDA_VISIBLE_DEVICES"] = ",".join([str(i) for i in range(nproc_per_node)])
        # 设置每个GPU的内存限制
        for i in range(nproc_per_node):
            os.environ[f"CUDA_VISIBLE_DEVICES_{i}"] = str(i)
            # 设置每个GPU的内存限制为20GB，留出4GB给系统和其他进程
            os.environ[f"CUDA_VISIBLE_DEVICES_{i}_MEMORY_LIMIT"] = "20GB"
        
        # 保存当前的sys.argv
        old_argv = sys.argv.copy()
        
        # 重定向标准输出和标准错误
        import io
        from contextlib import redirect_stdout, redirect_stderr
        import traceback
        
        log_output = io.StringIO()
        with open(log_file_path, 'w') as log_file, redirect_stdout(log_output), redirect_stderr(log_output):
            try:
                # 设置新的sys.argv，模拟命令行调用
                sys.argv = ["torch.distributed.run"] + torchrun_args
                
                # 直接调用torchrun.main()
                import torch.distributed.run as torchrun
                exit_code = torchrun.main()
                
                # 将输出写入日志文件
                log_file.write(log_output.getvalue())
                
                if exit_code != 0:
                    print(f"分布式任务执行失败，退出码: {exit_code}")
                    return (f"Error: 分布式任务执行失败，退出码: {exit_code}，查看日志: {log_file_path}",)
            
            except Exception as e:
                # 将输出写入日志文件
                log_file.write(log_output.getvalue())
                log_file.write(f"\n执行异常: {str(e)}\n")
                traceback.print_exc(file=log_file)
                
                print(f"启动分布式任务异常: {str(e)}")
                return (f"Error: 启动分布式任务异常: {str(e)}，查看日志: {log_file_path}",)
            
            finally:
                # 恢复原始的sys.argv
                sys.argv = old_argv
        
        # 检查输出文件是否存在
        output_path = os.path.join(save_dir, output_name)
        if os.path.exists(output_path):
            print(f"分布式任务执行成功，输出文件: {output_path}")
            return (output_path,)
        else:
            print(f"分布式任务执行失败，未找到输出文件: {output_path}")
            return (f"Error: 分布式任务执行失败，未找到输出文件，查看日志: {log_file_path}",)

**Comment by lcax200000** [comment]
Yes it is a worker node, you can try running it, but I always get cuda OOM at 720p
